Victory candlestick chart

ebrillhart opened the pull request:
@boygirl 

A basic chart currently renders as desired when the demo is run. Scale and range are calculated correctly for x and y values, but getDomain and getData both still need to be modified for this type of dataset. 

TODO

- [x] Pass styles down to candle component through this.props.style
- [x] Set default colors for positive and negative candles in getDataStyles (?)
- [x] Modify getData to accept four y values
- [x] Modify getDomain to accept four y values
- [x] eliminate flatten redundancy
- [x] figure out padding issue
- [x] figure out why x values don't quite line up with x-axis tick values
- [x] make sure non-numerical x values work (dates in particular)
- [x] write description for candlecolors prop
- [x] Write tests
- [x] Write more robust demos

boygirl commented on demo/components/victory-candlestick-demo.js:
```diff
@@ -0,0 +1,81 @@
+/*global window:false */
+import React from "react";
+import { random, range } from "lodash";
+import {VictoryCandlestick} from "../../src/index";
```
Please add a demo that shows rendering candlestick with VictoryChart

boygirl commented on src/components/victory-candlestick/candle.js:
```diff
@@ -0,0 +1,68 @@
+  }
+
+  renderCandle() {
+    const width = VictoryCandlestick.defaultProps.width;
```
This is an antipattern. Pass width and padding down as props to candle rather than referencing VictoryCandlestick here. Also, you should never need to reference defaultProps directly. These will just always show up in props, and that gives the user a change to override the defaults with passed props.

ebrillhart commented:
@boygirl still not ready to merge, but I implemented rudimentary getDomain and getData methods for this chart and would love some feedback. still am having trouble wrapping it in VictoryChart because the padding variable is passed down to the candle component as an object, not a number.

boygirl commented on src/components/victory-candlestick/candle.js:
```diff
@@ -20,8 +18,8 @@ export default class Candle extends React.Component {
-    const width = VictoryCandlestick.defaultProps.width;
-    const padding = VictoryCandlestick.defaultProps.padding;
+    const width = this.props.width;
+    const padding = this.props.padding;
```
Does the padding here refer to the padding of the entire chart as it does in VictoryChart, or is this related to padding between individual candlesticks? If so, you should expect to pluck padding off of the style object. Leave the padding prop on `VictoryCandlestick` the same as it is on the other components. It's used to calculate range, and therefore scale, so it needs to match up.

boygirl commented on src/components/victory-candlestick/helper-methods.js:
```diff
@@ -48,22 +49,69 @@ export default {
+
+  getDomain(props, axis) {
+    const dataset = this.getData(props);
+    const allData = flatten(dataset).map((datum) => {
```
Shouldn't need to flatten twice.

boygirl commented on src/components/victory-candlestick/helper-methods.js:
```diff
@@ -0,0 +1,130 @@
+      const candleHeight = Math.abs(scale.y(datum.open) - scale.y(datum.close));
+      const y = scale.y(Math.max(datum.open, datum.close));
+      const size = this.getSize(datum, props, calculatedValues);
+      const dataStyle = Object.assign(this.getDataStyles(datum, style.data),
```
actually cleaner to pass `props` into `getDataStyles` and have this logic happen there.

boygirl commented:
Does the padding prop required by candle refer to the padding of the entire chart as it does in VictoryChart, or is this related to padding between individual candlesticks? If so, you should expect to pluck padding off of the style object. Leave the padding prop on VictoryCandlestick the same as it is on the other components. It's used to calculate range, and therefore scale, so it needs to match up.

ebrillhart commented:
@boygirl I think this is really close to done - I've added some tests, most of which pass, and some additional demos. I also added a method to generate fallback data if none is provided. However, I'm having issues with the event handler tests and with event handling on data. Neither of the event handling tests pass (even though events on labels are working fine) and I can't get events to work on data components in the demo - not totally clear to me why these aren't working.

ebrillhart commented:
@boygirl now click events are working great, but the final two tests are still failing - not sure how much more I need to modify them for this component

boygirl commented on test/client/spec/components/victory-candlestick/victory-candlestick.spec.js:
```diff
@@ -0,0 +1,149 @@
+      const Labels = wrapper.find(VictoryLabel);
+      Labels.forEach((node, index) => {
+        node.childAt(0).simulate("click");
+        expect(clickHandler).called;
```
does just `expect(clickHandler).called` pass? what errors are you seeing?

ebrillhart replied:
I'm getting AssertionError: expected false to equal true for
expect(clickHandler.called).to.equal(true);

ebrillhart replied:
with your change, that now passes but getting the error

attaches an event to data
     components/victory-candlestick event handling
     TypeError: 'undefined' is not an object (evaluating 'clickHandler.args[index][1]')

for the test

expect(omit(clickHandler.args[index][1], ["events", "key"]))
          .to.eql(omit(initialProps, ["events", "key"]));

boygirl replied:
Let's pair tomorrow on some test debugging. I think it will go easier in person.

ebrillhart commented:
Oooh okay now only the second to last test fails (attaches an event to data)

boygirl commented on test/client/spec/components/victory-candlestick/victory-candlestick.spec.js:
```diff
@@ -0,0 +1,149 @@
+        node.simulate("click");
+        expect(clickHandler.called).to.equal(true);
+        // the first argument is the standard evt object
+        expect(omit(clickHandler.args[index][1], ["events", "key"]))
```
Try commenting out the second two expect statements and see if the clickHandler is even called. After that I would try `console.log`ing the clickHandler.args[index]. deep `eql` is pretty unforgiving.

ebrillhart replied:
if I have expect(clickHandler.called) it passes, if I have expect(clickHandler).called it does not

boygirl commented:
- [ ] default data

ebrillhart commented:
@boygirl realized I wasn't sure if you meant a default data set for the demo, or fallback data for the chart itself if no data was added. I already had fallback data, but I added a dataset to the demo file.

boygirl commented on src/components/victory-candlestick/helper-methods.js:
```diff
@@ -0,0 +1,156 @@
+    const max = Math.max(...domain);
+    const values = Array(...Array(samples)).map((val, index) => {
+      const v = (max / samples) * index + Math.min(...domain);
+      return { x: v, open: v, close: v + 2, low: v - 3, high: v + 5 };
```
Oh I see it now. This is your default data. I don't like having this hard coded into a method. I'm okay with the default data existing as a defaultProp as it does in `VictoryBar`. That way it will get documented in the props table.

boygirl commented:
So close! I would like you to add data accessors for x, high, low, open and close, but you can do that in a separate issue if you prefer.

ebrillhart commented:
@boygirl woo hoo! I'll add accessors now and then do theme support as a separate issue

ebrillhart commented:
so the issue with the accessors seems to exist only in the tests somehow. if I console log y to the browser console it's an array of integers (and all the demos run great), but if I console log it in the test suite it's an array of objects.

boygirl commented on src/components/victory-candlestick/victory-candlestick.js:
```diff
@@ -312,11 +390,16 @@ export default class VictoryCandlestick extends React.Component {
+    close: "close",
+    high: "high",
+    low: "low",
     y: "y",
```
Is this y accessor being used?

boygirl commented on test/client/spec/components/victory-candlestick/victory-candlestick.spec.js:
```diff
@@ -0,0 +1,151 @@
+    it("renders points for deeply-nested data", () => {
+      const data = range(40).map((i) => ({a: {b: [{x: i, y: i}]}}));
+      const wrapper = shallow(
+        <VictoryCandlestick data={data} x="a.b[0].x" y="a.b[0].y"/>
```
I would expect this spec to fail since you don't have high / low / open / close defined in your data or defined as accessors

boygirl commented on test/client/spec/components/victory-candlestick/victory-candlestick.spec.js:
```diff
@@ -0,0 +1,151 @@
+    it("renders data values with null accessor", () => {
+      const data = range(30);
+      const wrapper = shallow(
+        <VictoryCandlestick data={data} x={null} y={null}/>
```
I would expect this spec to fail since data will not be in a format the candlestick is expecting

ebrillhart commented:
@boygirl you're right about those tests - and those I've fixed with correctly formatted data. what I am confused about not passing are the tests for the helper methods, since their output is correct in the browser and demos but not in the test suite.

boygirl commented:
@ebrillhart y incorrect in your tests because you are passing y in with the data, and you are not passing in any open / close / high / low data or accessors. So those specs aren't actually testing your accessors. Still based on how your getData method is written, I would expect y to be `[]`. It might be helpful to write a few demos that make use of your new accessors and then come back and write the specs.
